### ISSUE TITLE
Add a slot system

### ISSUE THREAD

**Issue by LemADEC** [LemADEC/WarpDrive#34]
To help encourage player building choices, we need another variable than just power.
Size is currently used in the mod with multiblocks, but those are fairly complicated to implement and doesn't solve other mods.
Instead, we could implement a 'slot' system:
- bigger ship or better materials gets more slots,
- blocks on boards consume slots
- in-game tips remind on slot usage
- when too much slots are used, random extra-slots are disabled

This is add other interesting side effects:
- encourage players to keep the Warp core online, hence visible on radar
- progressively disable the ship as it loose 'volume'

**Comment by TheNumenorean** [comment]
This does not make sense to me. Slots in what?

**Comment by LemADEC** [comment]
Slots are used to place shield, weapon, power source, or engine components.
The idea is to encourage ship variety and progression, instead of a generic 'do everything' ship.
This also help counter the 200 Ultimate solar panels ship hull.

**Comment by lellasone** [comment]
Although the goal of encouraging diverse building choices is an admirable one, this feels like the wrong way to go about it. Right now you have a situation where players are exposed to diverse and complex emergent decisions based on the mods that they run. 

Implementing a slot system would needlessly restrict player choice and would encourage people to optimize the slots rather than figuring out what makes for a good ship in-game. 

As for the 200 ultimate solar ships (which I also find frustrating) those are just as valid a choice as any other for a player to make. Just a very poor one should that ship ever find itself in conflict (where the opponent would almost certainly use ICBM to force night and the costs involved would make risk taking Iffy)

If you want to encourage player decisions a much better route would be greater mod-integration and more complex mechanics. (for example, players could be encouraged to keep there warp cores online by having them lose their energy when disabled,). 

When you create mechanics that can only be solved using one mod's items and blocks that serves to restrict player choice and immersion rather than encouraging it.

**Comment by LemADEC** [comment]
I agree that slots by themselves won't work with a minecraft free-formed ship. We need another way of progression. Maybe requiring new materials to pass a certain ship size and encourage multi-blocks instead of 'all in one'.

Covering your space ship with solar panels makes perfect sense from immersion point of view. The issue is that those 200 Ultimate hybrids were killing tps on our dedicated box. Night won't work in space, since we have permanent day currently. On the other end, covering with opaque blocks would work in most cases, so we would need a umbrella deploying weapon, some kind of goo maybe?

As for greater mod integration, we're fairly extensive today, not quite sure sure which other one you've in mind here?
Indeed, a mainframe or ship core should always be linked for your on-board machines for them to work.

**Comment by LemADEC** [comment]
A weight system would make more sense with other mods that keep on compacting more storage in a single blocks.

**Comment by lellasone** [comment]
What do you mean by a weight system if I might ask?

**Comment by LemADEC** [comment]
Currently, the number of blocks in the ship is counted as its mass. The latest is used to enable hyperspace jump and compute energy costs.
The idea here is to count the content of containers and entities too to compute the ship mass.

This way, you can still have very compact storage with mega chests, nested containers or an AE system aboard, but still have impact on the ship movement (slower or no movement due to lack mass overload). Interestingly, you could use that effect to reach hyperspace with a smaller ship volume.

**Comment by CaptainMegavolt** [comment]
Well, in the case of compacting storage, that makes sense, but with applied
energistics, the items are supposedly converted into energy, which
shouldn't cause a noticeable net mass gain.

**Comment by LemADEC** [comment]
if AE is merely converting to energy, players would be able to duplicate items indefinitely from a single sample and more energy, like EMC based.
On the other end, it could be a different form of energy, one that isn't compatible with Warp fields and requires some form of stabilizer block.
Another option would be to give a different weight factor depending on the technology used for 'storage', giving the choice for players.

**Comment by CaptainMegavolt** [comment]
All very good ideas! As always, Mr. LemADEC, you know the way when it comes
to your mod. :)

**Comment by CaptainMegavolt** [comment]
I appreciate your taking my suggestions under consideration.

**Comment by Sasszem** [comment]
I have a basic idea instead of this 'slot system':
Add a basic multiblock hull, then attachable modules, maybe multiblock modules connected by cables. A hull can be connected to a limitted number of extensions, e.g a cable can connect two modues, so max modules are limitted by the num of connectors in hull. This would make a huge engine section on all ships.

All connectable modules should be multiblocks, like LESU from GT,
Modules looks like something like this:
-Capacitor,  increases capacity
-Shield generator: (re)generates shield blocks. Regenerating rate can be incresed by connecting a capacitor to it
-Cloaking core: time to generate field can be decreased by a capacitor
-Phaser bank: (replace current "laser medium"), stores phaser energy, capacity can be increased (do I need to tell, how?), connect to phaser cannons by special cable
-Warp coil? Decrease warp cooldown time

Also, Photon torpedoes.

Phasers destroy minimal amount of shield blocks, so they regenerate them => drains shield power. If you've got a slow shield, it simply drills in it.
Torpedoes explode, destroy large sections of shield, so they can make a hole if the generator is slow.
Also, phasers are precise, torpedoes are destructive.

Each shield should have a freq, including cloaking ones. A phaser or a torpedo with the same freq can go trougth them, so you don't need to know the owner of a beam. This would make the possibility to beat a stronger ship by good script(like change the phaser freq every shot then monitor the resoults). A camera with the same freq with the cloaking should see througth it.

**Comment by LemADEC** [comment]
Indeed, the slot system is about which blocks are present in your ship to unlock features or new blocks. My current idea is to just count them in the ship for unlocking (i.e. you need that many tier 2 structural blocks to have a ship that big).
Each structural block can be upgraded to be a wiring point where other mods can input/output power from. Those connections needs to form a wire to work properly.

I've never tried the GT LESU, it's an interesting idea that we could reuse here.

I'll probably add the shield system prior to the slots.
